results-processor/processor.py

```diff
@@ -18,200 +21,282 @@
+    def known_extension(path):
+        """Returns the extension of the path if known, otherwise None."""
+        EXT = ('.json.gz', '.txt.gz', '.gz', '.zip', '.json', '.txt')
+        for e in EXT:
+            if path.endswith(e):
+                return e
+        return None
+
+    def _download_gcs(self, gcs):
+        assert gcs.startswith('gs://')
+        ext = self.known_extension(gcs)
+        fd, path = tempfile.mkstemp(suffix=ext, dir=self._temp_dir)
+        os.close(fd)
+        # gsutil will log itself.
+        gsutil.copy(gcs, path)
+        return path
+
+    def _download_http(self, url):
+        assert url.startswith('http://') or url.startswith('https://')
+        _log.debug("Downloading %s", url)
+        r = requests.get(url, stream=True)
+        ext = (self.known_extension(r.header['Content-Disposition']) or
+               self.known_extension(url))
+        fd, path = tempfile.mkstemp(suffix=ext, dir=self._temp_dir)
+        with os.fdopen(fd, mode='wb') as f:
+            for chunk in r.iter_content(chunk_size=512*1024):
+                f.write(chunk)
+        os.close(fd)
+        return path
+
+    def _download_single(self, uri):
+        if uri.startswith('gs://'):
+            return self._download_gcs(uri)
+        return self._download_http(uri)
+
+    def _download_azure(self, azure_url):
+        artifact = self._download_http(azure_url)
+        with zipfile.ZipFile(artifact, mode='r') as z:
+            for f in z.infolist():
+                if f.is_dir():
+                    continue
+                path = z.extract(f, path=self._temp_dir)
+                if re.match(r'wpt_report.*\.json$', f.filename):
+                    self.results.append(path)
+                if re.match(r'wpt_screenshot.*\.txt$', f.filename):
+                    self.screenshots.append(path)
+
+    def download(self, results, screenshots, azure_url):
+        """Downloads all necessary inputs.
+
+        Args:
+            results: A list of results URIs (gs:// or https?://).
+            screenshots: A list of screenshots URIs (gs:// or https?://).
+            azure_url: A HTTP URL to an Azure build artifact.
+        """
+        if azure_url:
+            assert not results
+            assert not screenshots
+            self._download_azure(azure_url)
+            return
+        self.results = [self._download_single(i) for i in results]
+        self.screenshots = [self._download_single(i) for i in screenshots]
+
+    def load_report(self):
+        """Loads and merges all downloaded results."""
+        for r in self.results:
+            self.report.load_file(r)
+
+    def upload_raw(self):
+        """Uploads the merged raw JSON report to GCS."""
+        with tempfile.NamedTemporaryFile(
+                suffix='.json.gz', dir=self._temp_dir) as temp:
+            self.report.serialize_gzip(temp.name)
+            gsutil.copy(temp.name, self.raw_results_gs_url, gzipped=True)
+
+    def upload_split(self):
```

Comment:
Not reviewing this since it's in the refactoring commit and is similar to old code, even if this name is new.

Reply:
Yes this (as well as `upload_raw`) are indeed pure refactoring.